Source/Core/Core/HW/WII_IPC.cpp

```diff
@@ -200,13 +219,63 @@ void RegisterMMIO(MMIO::Mapping* mmio, u32 base)
                    // TODO: AVE, SLOT_LED
                  }));
   mmio->Register(base | GPIOB_DIR, MMIO::DirectRead<u32>(&gpio_dir.m_hex),
-                 MMIO::DirectWrite<u32>(&gpio_dir.m_hex));
+                 MMIO::ComplexWrite<u32>([](u32, u32 val) {
+                   gpio_dir.m_hex = (val & gpio_owner.m_hex) | (gpio_dir.m_hex & ~gpio_owner.m_hex);
+                 }));
   mmio->Register(base | GPIOB_IN, MMIO::ComplexRead<u32>([](u32) {
                    Common::Flags<GPIO> gpio_in;
                    gpio_in[GPIO::SLOT_IN] = DVDInterface::IsDiscInside();
                    return gpio_in.m_hex;
                  }),
                  MMIO::Nop<u32>());
+  // Starlet GPIO registers, not normally accessible by PPC (but they can be depending on how
+  // AHBPROT is set up).  We just always allow access, since some homebrew uses them.
+
+  // Note from WiiBrew: When switching owners, copying of the data is not necessary. For example, if
+  // pin 0 has certain configuration in the HW_GPIO registers, and that bit is then set in the
+  // HW_GPIO_OWNER register, those settings will immediately be visible in the HW_GPIOB registers.
+  // There is only one set of data registers, and the HW_GPIO_OWNER register just controls the
+  // access that the HW_GPIOB registers have to that data.
+  // Also: The HW_GPIO registers always have read access to all pins, but any writes (changes) must
+  // go through the HW_GPIOB registers if the corresponding bit is set in the HW_GPIO_OWNER
+  // register.
+  mmio->Register(base | GPIO_OUT, MMIO::DirectRead<u32>(&g_gpio_out.m_hex),
```

Comment:
How does this even work, since `base | 0xc0` is the same for both `GPIO_OUT` and `GPIOB_OUT` - does the 2nd one simply overwrite the 1st MMIO handler, or do _both_ run?

...unless the enum values up top are incorrect, and it's just that the last registered handler wins (which are those three)?

Reply:
Yep, the enum values are wrong (they should be e0/e4/e8).  It looks like it's the most recently registered handler that wins, too.  Will fix...

Reply:
Interresting that it does work though, perhaps the `GPIOB_*` ones are also slightly incorrect (enough to make a difference in some cases, but not enough to cause problems at large)?

Initially I wanted to comment on the enum values themselves, but then figured there might be some magic happening in the MMIO registration (so I put it down here)

Reply:
The GPIOs are only used for a few things -- the only things they handle are ejecting the disc and powering the sensor bar... and it turns out there was a big issue with this change already: the power sensor bar bit wasn't being set, so the cursor was always hidden.  Toggling power for the sensor bar is used in two places (blinking the sensor bar when adjusting sensitivity in the Wii menu, and pretending to be a TV remote for [the television friend channel](https://wiki.dolphin-emu.org/index.php?title=Television_Friend_Channel)).  (The eject function is either not used at all, or only used as an alternative path for the system menu where you boot, eject the disc, and then shut down; it's not something that dolphin can trigger normally.)  Thanks for catching this; it really would have broken things.